File: apps/server/src/ports/driven/services/id-generator.ts
// src/ports/driven/services/id-generator.ts

import { Context } from 'effect'
import type { Effect } from 'effect/Effect'

import type { ProductId } from '../../../domain/pilot'

// ============================================
// ID GENERATOR
// ============================================

export interface IdGeneratorService {
  readonly generateProductId: () => Effect<ProductId>
  readonly generateCorrelationId: () => Effect<string>
}

export class IdGenerator extends Context.Tag('IdGenerator')<IdGenerator, IdGeneratorService>() {}

File: apps/server-java/src/main/java/com/maisonamane/port/driven/service/IdGenerator.java
package com.maisonamane.port.driven.service;

import com.maisonamane.domain.pilot.valueobject.ProductId;
import com.maisonamane.domain.shared.CorrelationId;

/**
 * Interface for generating IDs.
 * Maps Effect-TS Context.Tag("IdGenerator") to Java interface.
 */
public interface IdGenerator {

    ProductId generateProductId();

    CorrelationId generateCorrelationId();
}
